File: tools/polyglot/neos_audit.pas
{ NeOS Profile Auditor in Free Pascal (FPC) }
program NeosAudit;

{$mode objfpc}{$H+}

uses
  SysUtils, Classes;

const
  RequiredFilesCount = 6;
  RequiredFiles: array[1..RequiredFilesCount] of string = (
    'profile/profiledef.sh',
    'profile/pacman.conf',
    'profile/grub/grub.cfg',
    'profile/syslinux/syslinux.cfg',
    'profile/packages.x86_64',
    'profile/airootfs/etc/pacman.d/neos-mirrorlist'
  );

var
  RootDir, FullPath, Line, PkgPath: string;
  I, PkgCount: Integer;
  PkgList: TStringList;
  PkgFile: TextFile;

begin
  if ParamCount >= 1 then
    RootDir := ParamStr(1)
  else
    RootDir := '.';

  WriteLn(#27'[1;36m[Pascal::Audit]'#27'[0m Auditing NeOS profile structure at: ', RootDir);

  for I := 1 to RequiredFilesCount do
  begin
    FullPath := IncludeTrailingPathDelimiter(RootDir) + RequiredFiles[I];
    if not FileExists(FullPath) then
    begin
      WriteLn(StdErr, '❌ Missing required profile file: ', FullPath);
      Halt(1);
    end;
  end;

  PkgPath := IncludeTrailingPathDelimiter(RootDir) + 'profile/packages.x86_64';
  if not FileExists(PkgPath) then
  begin
    WriteLn(StdErr, '❌ packages.x86_64 missing at: ', PkgPath);
    Halt(1);
  end;

  PkgList := TStringList.Create;
  try
    AssignFile(PkgFile, PkgPath);
    Reset(PkgFile);
    PkgCount := 0;
    while not Eof(PkgFile) do
    begin
      ReadLn(PkgFile, Line);
      Line := Trim(Line);
      if (Line <> '') and (Line[1] <> '#') then
      begin
        if PkgList.IndexOf(Line) >= 0 then
        begin
          WriteLn(StdErr, '❌ Duplicate package entry found in Pascal audit: ', Line);
          CloseFile(PkgFile);
          Halt(1);
        end;
        PkgList.Add(Line);
        Inc(PkgCount);
      end;
    end;
    CloseFile(PkgFile);
    WriteLn(#27'[1;32m✓ Free Pascal Audit Passed! Verified ', PkgCount, ' unique packages.'#27'[0m');
  finally
    PkgList.Free;
  end;
end.

File: tools/polyglot/NeosAudit.hs
-- NeOS Profile Auditor in Haskell (GHC)
module Main where

import System.Directory (doesFileExist)
import System.Environment (getArgs)
import System.Exit (exitFailure, exitSuccess)
import System.FilePath ((</>))
import Data.List (nub)

requiredFiles :: [FilePath]
requiredFiles =
  [ "profile/profiledef.sh"
  , "profile/pacman.conf"
  , "profile/grub/grub.cfg"
  , "profile/syslinux/syslinux.cfg"
  , "profile/packages.x86_64"
  , "profile/airootfs/etc/pacman.d/neos-mirrorlist"
  ]

main :: IO ()
main = do
  args <- getArgs
  let root = if null args then "." else head args
  putStrLn "\ESC[1;36m[Haskell::Audit]\ESC[0m Auditing NeOS profile invariants..."

  -- Check required files
  missing <- filterM' (\f -> not <$> doesFileExist (root </> f)) requiredFiles
  if not (null missing)
    then do
      putStrLn $ "❌ Missing required files: " ++ show missing
      exitFailure
    else do
      -- Parse package list
      content <- readFile (root </> "profile/packages.x86_64")
      let rawPkgs = filter (not . null) . map strip . lines $ content
      let pkgs = filter (not . isComment) rawPkgs
      if length pkgs /= length (nub pkgs)
        then do
          putStrLn "❌ Duplicate package entries found in Haskell audit"
          exitFailure
        else do
          putStrLn $ "\ESC[1;32m✓ Haskell Audit Passed! Verified " ++ show (length pkgs) ++ " packages.\ESC[0m"
          exitSuccess
  where
    strip = reverse . dropWhile (`elem` " \t\r\n") . reverse . dropWhile (`elem` " \t\r\n")
    isComment ('#':_) = True
    isComment _       = False

filterM' :: Monad m => (a -> m Bool) -> [a] -> m [a]
filterM' _ [] = return []
filterM' p (x:xs) = do
  flg <- p x
  ys <- filterM' p xs
  return $ if flg then x:ys else ys
